File: ruby/test/scl_parser_test.rb
require_relative "test_helper"

class SCLParserTest < Minitest::Test
  def test_loads_simple_configuration
    text = <<~SCL
      foo :: num { 42 }
      flag :: bool { true }
    SCL

    assert_equal({"foo" => 42, "flag" => true}, SCL.loads(text))
  end

  def test_dumps_round_trips_with_loads
    data = {"title" => "example", "count" => 3, "enabled" => false}

    serialized = SCL.dumps(data)

    assert serialized.end_with?("\n")
    assert_equal(data, SCL.loads(serialized))
  end

  def test_loads_raises_for_invalid_number
    text = <<~SCL
      foo :: num { bar }
    SCL

    error = assert_raises(SCL::SyntaxError) { SCL.loads(text) }
    assert_includes(error.message, ":number")
  end

  def test_load_reads_from_scl_file
    path = File.expand_path("fixtures/sample.scl", __dir__)
    assert File.exist?(path), "fixture is missing"

    result = SCL.load(path)
    expected = {"title" => "demo", "values" => [10, 20], "active" => false}

    assert_equal(expected, result)
  end
end

File: elixir/scl_parser/test/scl_parser_test.exs
defmodule SCLParserTest do
  use ExUnit.Case, async: true

  alias SCL.Errors.SyntaxError

  describe "loads/1" do
    test "parses a simple configuration" do
      text = """
      foo :: num { 42 }
      flag :: bool { true }
      """

      assert SCL.loads(text) == %{"foo" => 42, "flag" => true}
    end

    test "raises syntax error on invalid numeric value" do
      text = """
      foo :: num { bar }
      """

      assert_raise SyntaxError, ~r/Expected :number/, fn -> SCL.loads(text) end
    end
  end

  describe "dumps/1" do
    test "round-trips through loads/1" do
      data = %{"title" => "example", "count" => 3, "enabled" => false}

      serialized = SCL.dumps(data)

      assert String.ends_with?(serialized, "\n")
      assert SCL.loads(serialized) == data
    end
  end

  describe "load/1" do
    test "loads configuration from a .scl file" do
      path = Path.expand("fixtures/sample.scl", __DIR__)
      assert File.exists?(path)

      assert SCL.load(path) == %{
               "name" => "demo",
               "items" => [1, 2, 3],
               "enabled" => true
             }
    end
  end
end


